In [1]:
import pandas as pd

### Funnel for Cool T-Shirts Inc.

Cool T-Shirts Inc. has asked me to analyze data on visits to their website. My job is to build a funnel, which is a description of how many people continue to the next step of a multi-step process.

In this case, our funnel is going to describe the following process:

A user visits CoolTShirts.com

A user adds a t-shirt to their cart

A user clicks “checkout”

A user actually purchases a t-shirt

In [2]:
# Loading and reading data from the csv files 

In [3]:
cart = pd.read_csv(r'C:\Users\p\Documents\Data Manipulation  with Pandas\cart.csv', encoding = "ISO-8859-1")

In [4]:
checkout = pd.read_csv(r'C:\Users\p\Documents\Data Manipulation  with Pandas\checkout.csv', encoding = "ISO-8859-1")

In [5]:
purchase = pd.read_csv(r'C:\Users\p\Documents\Data Manipulation  with Pandas\purchase.csv', encoding = "ISO-8859-1")

In [6]:
visits = pd.read_csv(r'C:\Users\p\Documents\Data Manipulation  with Pandas\visits.csv', encoding = "ISO-8859-1")

Inspect the DataFrames using print and head:

visits lists all of the users who have visited the website

cart lists all of the users who have added a t-shirt to their cart

checkout lists all of the users who have started the checkout

purchase lists all of the users who have purchased a t-shirt

In [7]:
# Inspecting the cart DataFrame using print and head:

print(cart.head()) 

                                user_id        cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168  11/7/2017 20:45
1  4397f73f-1da3-4ab3-91af-762792e25973   5/27/2017 1:35
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663   3/4/2017 10:38
3  b594862a-36c5-47d5-b818-6e9512b939b3   9/27/2017 8:22
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974  7/26/2017 15:48


In [8]:
# Inspecting the checkout DataFrame using print and head:

print(checkout.head()) 

                                user_id    checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d   6/25/2017 9:29
1  4ac186f0-9954-4fea-8a27-c081e428e34e   4/7/2017 20:11
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d  7/13/2017 11:38
3  89fe330a-8966-4756-8f7c-3bdbcd47279a  4/20/2017 16:15
4  3ccdaf69-2d30-40de-b083-51372881aedd   1/8/2017 20:52


In [9]:
# Inspecting the purchase DataFrame using print and head:

print(purchase.head()) 

                                user_id     purchase_time
0  4b44ace4-2721-47a0-b24b-15fbfa2abf85    5/11/2017 4:25
1  02e684ae-a448-408f-a9ff-dcb4a5c99aac     9/5/2017 8:45
2  4b4bc391-749e-4b90-ab8f-4f6e3c84d6dc  11/20/2017 20:49
3  a5dbb25f-3c36-4103-9030-9f7c6241cd8d   1/22/2017 15:18
4  46a3186d-7f5a-4ab9-87af-84d05bfd4867   6/11/2017 11:32


In [10]:
# Inspecting the visits DataFrame using print and head:

print(visits.head())

                                user_id       visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188   4/7/2017 15:14
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  1/26/2017 14:24
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1   8/20/2017 8:23
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  11/4/2017 18:15
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2/27/2017 11:25


Combine visits and cart using a left merge.

In [11]:
# Combining visits and cart using a left merge to find customers who visited the website and added a t-shirt to their carts

visits_cart = pd.merge(visits, cart, how ='left')

visits_cart.head(10)

user_id       visit_time        cart_time
0  943647ef-3682-4750-a2e1-918ba6f16188   4/7/2017 15:14              NaN
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  1/26/2017 14:24  1/26/2017 14:44
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1   8/20/2017 8:23   8/20/2017 8:31
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  11/4/2017 18:15              NaN
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2/27/2017 11:25              NaN
5  7bc62113-d51d-4e6b-85e0-1cf29e3de74a   6/3/2017 20:05              NaN
6  6f22cd44-bc63-4449-a9af-2651859da2f9  3/18/2017 23:09              NaN
7  49a6399b-c095-4e42-92eb-af90abe2bb6a  1/10/2017 12:09              NaN
8  653c095d-fd74-40dd-ae3b-5a69a2613629  6/27/2017 19:21              NaN
9  3d9fdac1-aae3-4a01-b448-934d12047468   9/24/2017 6:43              NaN

How long is your merged DataFrame?

In [12]:
# finding the length of the merged DataFrame

visits_cart_rows = len(visits_cart)

visits_cart_rows

2000

How many of the timestamps are null for the column cart_time?

What do these null rows mean?

In [13]:
# finding the number of timestamps that are null for the column cart_time

# these are people who visited the website but did not add a t-shirt to their carts

Null_cart_time = len(visits_cart[visits_cart.cart_time.isnull()])

Null_cart_time

1652

What percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart?
##### Note: 
To calculate percentages, it will be helpful to turn either the numerator or the denominator into a float, by using float(), with the number to convert passed in as input. Otherwise, Python will use integer division, which truncates decimal points.

In [14]:
# percent of users who visited Cool T-Shirts Inc. but did not place a t-shirt in their cart

Null_cart_time_percent = (float(Null_cart_time) / (visits_cart_rows)) * 100

Null_cart_time_percent

82.6

Repeat the left merge for cart and checkout and count null values. What percentage of users put items in their cart, but did not proceed to checkout?

In [15]:
# Combining cart and checkout using a left merge to find customers who added a t-shirt to their carts and started checkout 

cart_checkout = pd.merge(cart, checkout, how = 'left')

cart_checkout.head(10)

user_id        cart_time    checkout_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168  11/7/2017 20:45  11/7/2017 21:14
1  2be90e7c-9cca-44e0-bcc5-124b945ff168  11/7/2017 20:45  11/7/2017 20:50
2  2be90e7c-9cca-44e0-bcc5-124b945ff168  11/7/2017 20:45  11/7/2017 21:11
3  4397f73f-1da3-4ab3-91af-762792e25973   5/27/2017 1:35              NaN
4  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663   3/4/2017 10:38   3/4/2017 11:04
5  b594862a-36c5-47d5-b818-6e9512b939b3   9/27/2017 8:22   9/27/2017 8:26
6  a68a16e2-94f0-4ce8-8ce3-784af0bbb974  7/26/2017 15:48              NaN
7  0c998868-9961-490b-8ec2-9981ba667027  11/27/2017 8:54  11/27/2017 9:18
8  f783c680-1d9a-437d-9f45-7827299b78fa   6/25/2017 8:08   6/25/2017 8:28
9  f783c680-1d9a-437d-9f45-7827299b78fa   6/25/2017 8:08   6/25/2017 8:27

In [16]:
# number of rows of of people who added a t-shirt to cart but did not begin checkout

Null_cart_checkout = len(cart_checkout[cart_checkout.checkout_time.isnull()])

Null_cart_checkout

122

In [17]:
# length of cart_checkout showing total no. rows of people who added a t-shirt to cart & did or did not begin checkout

cart_checkout_rows = len(cart_checkout)

cart_checkout_rows

482

What percentage of users put items in their cart, but did not proceed to checkout?

In [18]:
# percentage of users who put items in their cart, but did not proceed to checkout

Null_cart_checkout_percent = (float(Null_cart_checkout) / cart_checkout_rows) * 100

round(Null_cart_checkout_percent,2)

25.31

Merge all four steps of the funnel, in order, using a series of left merges. Save the results to the variable all_data.

Examine the result using print and head.

In [19]:
all_data = visits.merge(cart, how = 'left').merge(checkout, how = 'left').merge(purchase, how = 'left')

all_data.head(10)

user_id       visit_time        cart_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188   4/7/2017 15:14              NaN   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  1/26/2017 14:24  1/26/2017 14:44   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1   8/20/2017 8:23   8/20/2017 8:31   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  11/4/2017 18:15              NaN   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2/27/2017 11:25              NaN   
5  7bc62113-d51d-4e6b-85e0-1cf29e3de74a   6/3/2017 20:05              NaN   
6  6f22cd44-bc63-4449-a9af-2651859da2f9  3/18/2017 23:09              NaN   
7  49a6399b-c095-4e42-92eb-af90abe2bb6a  1/10/2017 12:09              NaN   
8  653c095d-fd74-40dd-ae3b-5a69a2613629  6/27/2017 19:21              NaN   
9  3d9fdac1-aae3-4a01-b448-934d12047468   9/24/2017 6:43              NaN   

     checkout_time    purchase_time  
0              NaN              NaN  
1  1/26/2017 14:54  1/26/2017 15:08  
2              NaN              NaN  
3              NaN              NaN  
4              NaN              NaN  
5              NaN              NaN  
6              NaN              NaN  
7              NaN              NaN  
8              NaN              NaN  
9              NaN              NaN

What percentage of users proceeded to checkout, and purchased a t-shirt?

In [20]:
checkout_purchase = pd.merge(checkout, purchase, how = 'left')

checkout_purchase.head()

user_id    checkout_time   purchase_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d   6/25/2017 9:29  6/25/2017 9:49
1  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d   6/25/2017 9:29  6/25/2017 9:56
2  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d   6/25/2017 9:29  6/25/2017 9:57
3  4ac186f0-9954-4fea-8a27-c081e428e34e   4/7/2017 20:11             NaN
4  3c9c78a7-124a-4b77-8d2e-e1926e011e7d  7/13/2017 11:38             NaN

In [21]:
checkout_purchase_rows = len(checkout_purchase)

checkout_purchase_rows

598

In [22]:
# showing the clients who completed checkout and purchased a t-shirt 

Made_checkout_purchase = checkout_purchase[~checkout_purchase.purchase_time.isnull()]

Made_checkout_purchase.head(5)

user_id    checkout_time    purchase_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d   6/25/2017 9:29   6/25/2017 9:49
1  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d   6/25/2017 9:29   6/25/2017 9:56
2  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d   6/25/2017 9:29   6/25/2017 9:57
5  89fe330a-8966-4756-8f7c-3bdbcd47279a  4/20/2017 16:15  4/20/2017 16:15
6  3ccdaf69-2d30-40de-b083-51372881aedd   1/8/2017 20:52   1/8/2017 21:02

In [23]:
# finding the total number of people who completed checkout and made a purchase 

Made_checkout_purchase = len(checkout_purchase[~checkout_purchase.purchase_time.isnull()])

Made_checkout_purchase 

497

In [24]:
# percentage of users proceeded to checkout, and purchased a t-shirt

Made_checkout_purchase_percent = (float(Made_checkout_purchase) / checkout_purchase_rows) * 100

round(Made_checkout_purchase_percent, 2)

83.11

Which step of the funnel is weakest (i.e., has the highest percentage of users not completing it)?

How might Cool T-Shirts Inc. change their website to fix this problem?

##### Answer:
The first step of the funnel is weakest since it has the highest percentage of users **(82.6)** not completing it.

Cool T-Shirts Inc. can change their website to fix this problem by enticing visitors with 'goodies' such as product discounts, free gifts or samples for visitors who proceed and complete the whole funnel process.